# <table align="center">
  <tr >
    <td>
      <img src="https://drive.usercontent.google.com/download?id=1PW43TkPAGlUge80Ucid2c_OeL-6zdYh9" align="left" />
    </td>
    <td>
      <h1>
        TRƯỜNG ĐẠI HỌC NHA TRANG <br><b>KHOA CÔNG NGHỆ THÔNG TIN</b>
        </h1>
    </td>
  </tr>
</table>

<h1><b> THỰC HÀNH LẬP TRÌNH PYTHON</b></h1>

# Bài thực hành 1: Python cơ bản

---



## Bài 1.1: Chuẩn bị môi trường lập trình
### Mục đích:
    - Biết thiết lập môi trường lập trình Python trực tuyến.
    - Viết & thực thi một vài chương trình đơn giản.
  
### Lập trình Python với Google Colaboratory
  - Ưu điểm
    - Môi trường lập trình tương tác, trực tuyến trên giao diện web.
    - Không cần cài đặt môi trường lập trình, có sẵn các thư viện thông dụng.
    - Tích hợp sẵn AI trợ giúp, phù hợp cho việc học lập trình.
  - Hạn chế
    - Máy tính cần kết nối Internet.
    - Không đầy đủ tính năng như các phần mềm lập trình Python cài đặt trên máy tính (như PyCharm, Spyder...)

> **Chú ý:**  Ngoài Google Colab, có một số môi trường lập trình Python trực tuyến miễn phí khác như: [Online Python](https://www.online-python.com/), [W2SchoolPython Online Compiler](https://www.w3schools.com/python/python_compiler.asp).


### Chương trình Python đầu tiên  
- [Mở môi trường lập trình Google Colaboratory](https://colab.research.google.com/).
- Click **"New Notebook"** để tạo một tập tin mã nguồn mới. Đổi tên notebook thành "Bài thực hành 1" và lưu vào thư mục "Lập trình Python" trên Google Drive.
- Tại ô viết mã lệnh, nhập vào lệnh sau:

In [ ]:
# In ra câu chào "Hello Python coders!"
print("Hello Python coders!")

Hello Python coders!


- Click biểu tượng mũi tên bên trái ô lệnh, hoặc nhấn tổ hợp phím (Ctrl+Enter) để chạy đoạn lệnh.
- Quan sát kết quả trên màn hình.



---


## Bài 1.2: Chào Anh Tuấn, chào mừng bạn đến Đại học Nha Trang!

### Mục đích:
    - Sử dụng được lệnh nhập, xuất dữ liệu dạng chuỗi.
    - Thao tác với biến, biểu thức.

### Các bước thực hiện

- Chèn một ô mã nguồn vào bên dưới (click **+Code**) hoặc phím tắt **B**.
- Nhập đoạn lệnh sau: [link text](https://)

In [ ]:
# Viết chương trình mời nhập vào tên của một người, in ra lời chào người đó
# Lệnh đọc chuỗi ký tự từ bàn phím và đưa vào biến 'name'
name = input("Xin nhập tên của bạn: ")
# In ra lời chào người có tên vừa nhập
print("Chào " + name + ", chào mừng bạn đến Đại học Nha Trang!")

Xin nhập tên của bạn: Hưng
Chào Hưng, chào mừng bạn đến Đại học Nha Trang!


- Chạy đoạn lệnh trên, nhập vào một tên khi được yêu cầu và nhấn Enter. Quan sát kết quả trên màn hình.

## Bài 1.3: Tính chu vi, diện tích hình tròn

### Mục đích:
  - Sử dụng được lệnh nhập, xuất dữ liệu và chuyển đổi sang kiểu số.
  - Biết sử dụng thư viện các hàm có sẵn của Python.

### Các bước thực hiện
- Chèn một ô mã nguồn vào bên dưới.
- Nhập đoạn lệnh sau:

In [ ]:
# Lệnh đọc giá trị bán kính hình tròn
ban_kinh = float(input("Nhập bán kính = "))
# Tính chu vi hình tròn, biết hằng số PI gần bằng 3.1416
chu_vi = ban_kinh * 2 * 3.1416
# Tính diện tích hình tròn
dien_tich = ban_kinh ** 2 * 3.1416
# In ra diện tích hình tròn
print("Chu vi hình tròn = ", chu_vi)
# In ra diện tích hình tròn
print(f"Diện tích hình tròn = {dien_tich:.2f}")


Nhập bán kính = 2
Chu vi hình tròn =  12.5664
Diện tích hình tròn = 12.57


- Chạy đoạn code trên, nhập dữ liệu theo yêu cầu và quan sát kết quả.
- Giải thích dòng lệnh 2, hàm float() có tác dụng gì?
- Giải thích dòng lệnh 6.
- So sánh lệnh in ra chu vi (dòng 8) với lệnh in ra diện tích (dòng 10).
- Chạy lại đoạn code và cố ý nhập vào một giá trị bán kính < 0, quan sát kết quả. Dù bán kính < 0 nhưng máy tính vẫn tính & in ra chu vi & diện tích, như vậy có đúng logic chưa? Cách khắc phục lỗi này? (Gợi ý: dùng lệnh điều kiện if).
- Chạy lại đoạn code và cố ý nhập vào một giá trị không phải kiểu số (vd 123ABC), quan sát kết quả. Vì sao máy tính báo lỗi "ValueError"? Cách khắc phục?
- Hãy sửa lại đoạn code để in ra giá trị diện tích được làm tròn đến 2 chữ số phần thập phân.
- Hãy sử dụng hằng số PI từ thư viện có sẵn theo gợi ý sau để có kết quả chính xác hơn:
```python
# Lệnh nạp thư viện toán học của Python
import math

# In ra giá trị số PI
print(math.pi)
# Tính diện tích hình tròn
dien_tich = math.pi * ban_kinh ** 2
```


## Bài 1.4: Tính chỉ số BMI
BMI (Body Mass Index) là một chỉ số giúp đánh giá thể trạng của một người (thiếu cân, bình thường, thừa cân, béo phì) dựa vào cân nặng và chiều cao của họ. Chỉ số này được tính bằng công thức:
$$
  bmi = \frac{{weight}}{height^2}
$$
trong đó cân nặng (weight) (đơn vị kg) chia cho bình phương chiều cao (height) (đơn vị mét). BMI là công cụ sàng lọc phổ biến, giúp xác định nguy cơ sức khỏe liên quan đến cân nặng và là cơ sở để điều chỉnh chế độ ăn uống, sinh hoạt.

### Mục đích:
  - Sử dụng được lệnh nhập, xuất dữ liệu và chuyển đổi sang kiểu dữ liệu phù hợp.
  - Sử dụng câu lệnh điều khiển rẽ nhánh thích hợp.

### Các bước thực hiện
- Chèn một ô mã nguồn vào bên dưới.
- Nhập đoạn lệnh sau:

In [ ]:
# Viết chương trình tính chỉ số BMI

# Đọc giá trị cân nặng (kg)
weight = float(input("Nhập cân nặng (kg): "))
# Đọc giá trị chiều cao (mét)
height = float(input("Nhập chiều cao (mét): "))

# Tính chỉ số BMI
bmi = weight / (height ** 2)

# In ra chỉ số BMI
print(f"Chỉ số BMI của bạn là: {bmi:.2f}")

Nhập cân nặng (kg): 67
Nhập chiều cao (mét): 1.65
Chỉ số BMI của bạn là: 24.61


- Chạy đoạn code trên và theo dõi kết quả.
- Hãy bổ sung chương trình để in ra phân loại sức khỏe theo BMI dựa theo bảng phân loại sau:

| Chỉ số BMI (kg/m²)       | Phân loại                         | Mức độ nguy cơ đối với sức khỏe |
|---------------------------|-----------------------------------|----------------------------------|
| < 18,5                    | Thiếu cân                        | Thấp (nhưng có thể thiếu dinh dưỡng, loãng xương) |
| 18,5 – 22,9               | Bình thường                      | Trung bình                      |
| 23,0 – 24,9               | Thừa cân                         | Có nguy cơ                      |
| ≥ 25,0                | Béo phì                      | Cao                             |


In [ ]:
# Viết chương trình tính chỉ số BMI

# Đọc giá trị cân nặng (kg)
weight = float(input("Nhập cân nặng (kg): "))
# Đọc giá trị chiều cao (mét)
height = float(input("Nhập chiều cao (mét): "))

# Tính chỉ số BMI
bmi = weight / (height ** 2)

# In ra chỉ số BMI
print(f"Chỉ số BMI của bạn là: {bmi:.2f}")
# Đánh giá tình trạng cân nặng dựa trên BMI
if bmi < 18.5:
    print("Tình trạng: Thiếu cân")
elif bmi <= 22.9:
    print("Tình trạng: Cân nặng bình thường")
elif 23 <= bmi <= 24.9:
    print("Tình trạng: Thừa cân")
else:
    print("Tình trạng: Béo phì")

Nhập cân nặng (kg): 67
Nhập chiều cao (mét): 1.65
Chỉ số BMI của bạn là: 24.61
Tình trạng: Thừa cân


## Luyện tập

### Bài 1.5: Giải phương trình bậc hai $ax^2+bx+c=0$, với $a,b,c$ là các số thực bất kỳ nhập từ bàn phím.

### Bài 1.6: Nhập một năm dương lịch, tính xem năm đó có phải năm nhuận hay không. Biết năm nhuận là năm thỏa mãn một trong hai điều kiện:
- Chia hết 400, hoặc
- Chia hết 4 và không chia hết 100.

### Bài 1.7: Cho bảng giá điện sinh hoạt giả định như sau:
| Bậc | Mức sử dụng (kWh)     | Đơn giá (đồng/kWh) |
|-----|-------------------------|------------------|
| 1   | 0 – 50                  | 1750 |
| 2   | 51 – 100                | 2000 |
| 3   | >= 101                  | 2500 |

Viết chương trình đọc số kWh tiêu thụ, dựa vào bảng giá tính số tiền điện phải trả.

### Bài 1.8: Viết chương trình nhập các điểm thành phần môn học của một sinh viên: điểm quá trình, điểm thi giữa kỳ, điểm thi cuối kỳ.
- Tính điểm tổng kết học phần của sinh viên đó theo công thức
$$
    \text{Điểm tổng kết học phần = Điểm quá trình * 30% + Điểm giữa kỳ * 30% + Điểm cuối kỳ * 40%}
$$
- In thông báo "Đạt" nếu điểm tổng kết >= 5.0, ngược lại in thông báo "Không đạt"

### Bài 1.9: Một máy ATM có các loại tiền mệnh giá 500000, 200000, 100000, và 50000 VND. Một người cần rút số tiền M (VND). Hãy đưa ra phương án rút được số  tiền M sao cho số tờ tiền là ít nhất.

Ví dụ: Với số tiền cần rút là 950000000, phương án tối ưu là 1 tờ 500000 + 2 tờ 200000 + 1 tờ 50000 VND.

> **Gợi ý:** Sử dụng các toán tử trên số nguyên: phép chia lấy phần nguyên (//), lấy số dư (%)
> - Nhập số tiền M.
> - Kiểm tra M có thỏa mãn là số nguyên > 0 và chia hết cho mệnh giá nhỏ nhất (50000):
>     - Nếu thỏa mãn:
>        - Tính số tờ tiền theo mệnh giá cao nhất.
>        - Tính số tiền còn lại.
>        - Lần lượt đổi ra số tờ tiền cho các mệnh giá thấp hơn kế tiếp cho đến khi đủ số tiền cần rút.
>        - In ra kết quả (số lượng tờ tiền mỗi loại)
>    - Nếu không thỏa:
>        - Thông báo "Không rút được"

### Bài 1.10: ĐH Nha Trang lập kế hoạch thi cuối kỳ cho một lớp, dự kiến sử dụng ba phòng thi P1, P2, P3 với số sinh viên tương ứng là n1, n2, n3. Quy định mỗi bàn không bố trí quá hai chỗ ngồi. Hãy viết chương trình thực hiện các yêu cầu sau:
- Nhập số lượng sinh viên của mỗi phòng thi.
- Tính số lượng bàn tối thiểu cần bố trí cho mỗi phòng.
- Tính tổng số lượng bàn cần bố trí.

Ví dụ: Nếu xếp số SV của các phòng lần lượt là 35, 42, 39 thì số lượng bàn tối thiểu tương ứng mỗi phòng là 18, 21 và 20; tổng số bàn cần dùng là 59.
> **Gợi ý:** Sử dụng hàm ceil() trong thư viện toán học (math) của Python.




